### ISSUE TITLE
Rule Proposal: warning if assigning Record to interfaces with optionals

### ISSUE THREAD

**Issue by villasv** [typescript-eslint/typescript-eslint#2542]
Consider the following code:

```typescript
interface ObjWithPropFalsy {
    prop?: false;
}

const a: Record<string, boolean> = { prop: true };
const b: ObjWithPropFalsy = a;

if (b.prop) { // b.prop is of type false | undefined, but its value is true
    console.log("typing says it never happens") // <-- will print
} else {
    console.log("typing says it always happens")
}
```

This compiles fine, because `Record<string, boolean>` is assignable to `ObjWithPropFalsy` since `prop` is optional. I think this is pretty dangerous, and although sometimes it can be safely ignored, it should raise some eyebrows about type safety.

If I change the condition to `if (b.prop === true)` I even get a compilation error because there's no type overlap.

Might be worth it extending the rule to account not only `Record` but any index signatures? Not sure about how the general case works.

**Comment by bradzacher** [comment]
I would consider raising this against teh TS repo itself.

This is an inherit weakness in TS itself. The type of the record need not be `string, bool`, it could be `string, anything`.

Could make a lint rule for this, but IMO it would be much better to talk to the TS team about fixing this unsafety instead of relying on linting.

Example:

```ts
interface ObjWithPropFalsy {
    prop?: false;
}

const a: Record<string, string> = { prop: 'str' };
const b: ObjWithPropFalsy = a;

type T = (typeof b)['prop']; // === false | undefined;
typeof a.prop === 'string' // runtime
```

**Comment by villasv** [comment]
Backlink: https://github.com/typescript-eslint/typescript-eslint/issues/2542

**Comment by bradzacher** [comment]
Awesome!
I'll close this for now.
If the TS team comes back and decides this won't be fixed in TS, we can reopen and revisit this idea.
